distage/distage-core-api/src/main/scala/izumi/distage/model/definition/dsl/AbstractBindingDefDSL.scala

```diff
@@ -200,9 +207,15 @@ object AbstractBindingDefDSL {
           val newRef = SingletonBinding(key, ImplDef.ReferenceImpl(b.implementation.implType, b.key, weak = false), b.tags, pos)
           refs = newRef :: refs
         case Modify(providerMagnetModifier) =>
-          ???
-//          val newRef = SingletonBinding(key, ImplDef.ProviderImpl(b.implementation.implType, ???), b.tags, ???)
-//          refs = newRef :: refs
+          b.implementation match {
+            case ImplDef.ProviderImpl(implType, function) =>
+              val oldMagnet = ProviderMagnet(function)
+              val newProvider = providerMagnetModifier(oldMagnet).get
+              if (implType == newProvider.ret) {
+                b = b.withImplDef(ImplDef.ProviderImpl(newProvider.ret, newProvider))
```

Comment:
Why conditional update? We forbid incompatible types at API level & it's safe for the new type to be `<:` than less previous `implType`

Reply:
Yeah, you are right. I've added this check instead: `if (newProvider.ret <:< implType)`. Is this enough?